File: WordTarget/WordTargetCoreTests/WordDatabaseTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System.IO;
using static WordTargetCore.CommonTestObjects;

namespace WordTargetCore
{
    [TestClass]
    public class WordDatabaseTest
    {


        [TestMethod]
        public void PopulatingDatabase()
        {
            WordDatabase db = new WordDatabase();
            db.AddWord(cat);
            db.AddWords(new List<Word> { bat, baseball, sport, ports, port });
            Assert.IsTrue(db.GetAllWords().SetEquals(new HashSet<Word> { cat, bat, baseball, sport, ports, port }));
            HashSet<Link> batLinks = db.GetLinksFor(bat);
            Assert.AreEqual(1, batLinks.Count);
            Assert.IsTrue(batLinks.Contains(catBat.Reverse()));
            HashSet<Link> portsLinks = db.GetLinksFor(ports);
            Assert.AreEqual(2, portsLinks.Count);
            Assert.IsTrue(portsLinks.Contains(new Link(ports, sport, LinkType.Anagram)));
            Assert.IsTrue(portsLinks.Contains(new Link(ports, port, LinkType.OneLetterAddOrRemove)));
        }

        [TestMethod]
        public void SerializeDatabase()
        {
            string expected = @"Words:
BASEBALL, BAT, CAT, PORT, PORTS, SPORT

Links:
BAT CAT OneLetterChange
CAT BAT OneLetterChange
PORT PORTS OneLetterAddOrRemove
PORT SPORT OneLetterAddOrRemove
PORTS PORT OneLetterAddOrRemove
PORTS SPORT Anagram
SPORT PORT OneLetterAddOrRemove
SPORT PORTS Anagram
";
            StringWriter writer = new StringWriter();
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball, sport, ports, port });
            db.Serialize(writer);
            Assert.AreEqual(expected, writer.GetStringBuilder().ToString());
        }

        [TestMethod]
        public void DeserializeDatabase()
        {
            string expected = @"Words:
BASEBALL, BAT, CAT, PORT, PORTS, SPORT

Links:
BASEBALL BAT WordAssociation
BAT BASEBALL WordAssociation
BAT CAT OneLetterChange
CAT BAT OneLetterChange
PORT PORTS OneLetterAddOrRemove
PORT SPORT OneLetterAddOrRemove
PORTS PORT OneLetterAddOrRemove
PORTS SPORT Anagram
SPORT PORT OneLetterAddOrRemove
SPORT PORTS Anagram
";
            StringReader reader = new StringReader(expected);
            WordDatabase db = WordDatabase.Deserialize(reader);
            Assert.IsTrue(db.GetAllWords().SetEquals(new HashSet<Word> { baseball, bat, cat, sport, ports, port }));
            Assert.IsTrue(db.ContainsLink("BASEBALL", "BAT"));
            Assert.IsTrue(db.ContainsLink("BAT", "BASEBALL"));
            Assert.IsTrue(db.ContainsLink("PORTS", "SPORT"));
        }

        [TestMethod]
        public void FindChain()
        {
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball, sport, ports, port });
            db.AddLink(new Link(bat, baseball, LinkType.WordAssociation));
            db.AddLink(new Link(sport, baseball, LinkType.WordAssociation));
            Chain chain = db.FindChain(cat, port);
            Assert.AreEqual(5, chain.Count);
            Assert.AreEqual(catBat, chain[0]);
            Assert.AreEqual(batBaseball, chain[1]);
            Assert.AreEqual(baseballSport, chain[2]);
            Assert.AreEqual(sportPorts, chain[3]);
            Assert.AreEqual(portsPort, chain[4]);
        }

        [TestMethod]
        public void ImproperWordAdd()
        {
            WordDatabase db = new WordDatabase();
            db.AddWord(cat);
            Assert.IsTrue(db.GetAllWords().SetEquals(new HashSet<Word> { cat }));
            db.AddWord(cat);
            Assert.AreEqual(1, db.GetAllWords().Count);
            Assert.IsTrue(db.GetAllWords().SetEquals(new HashSet<Word> { cat }));
        }

        [TestMethod]
        public void ContainsWord()
        {
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball });
            Assert.IsTrue(db.ContainsWord(cat));
            Assert.IsTrue(db.ContainsWord("BAT"));
            Assert.IsTrue(db.ContainsWord(baseball));
            Assert.IsFalse(db.ContainsWord(ports));
            Assert.IsFalse(db.ContainsWord("SPORT"));
            Assert.IsFalse(db.ContainsWord(""));
        }

        [TestMethod]
        public void ContainsLink()
        {
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball, sport, ports, port });
            db.AddLink(new Link(bat, baseball, LinkType.WordAssociation));
            Assert.IsTrue(db.ContainsLink(cat, bat));
            Assert.IsTrue(db.ContainsLink(bat, cat));
            Assert.IsTrue(db.ContainsLink(bat, baseball));
            Assert.IsTrue(db.ContainsLink("SPORT", "PORTS"));
            Assert.IsFalse(db.ContainsLink(cat, sport));
            Assert.IsFalse(db.ContainsLink("CAT", "RAT"));
        }

        [TestMethod]
        public void GetLinkBetween()
        {
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball, sport, ports, port });
            db.AddLink(batBaseball);
            db.AddLink(baseballSport);
            Assert.AreEqual(catBat, db.GetLinkBetween(cat, bat));
            Assert.AreEqual(baseballSport, db.GetLinkBetween(baseball, sport));
            Assert.AreEqual(null, db.GetLinkBetween(cat, baseball));
            Assert.AreEqual(null, db.GetLinkBetween(car, cat));
            Assert.AreEqual(null, db.GetLinkBetween("", "CAT"));
        }

        [TestMethod]
        public void RemoveLink()
        {
            WordDatabase db = new WordDatabase();
            db.AddWords(new List<Word> { cat, bat, baseball, sport, ports, port });
            db.AddLink(batBaseball);
            db.AddLink(baseballSport);
            Assert.IsTrue(db.ContainsLink(sport, port));
            db.RemoveLink(sportPort);
            Assert.IsFalse(db.ContainsLink(sport, port));
            Assert.ThrowsException<System.Exception>(() => db.RemoveLink(portsParts));
        }
    }
}

File: word-target/word-target-core/src/test/java/it/carcassi/wordtarget/core/WordDatabaseTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.carcassi.wordtarget.core;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import static it.carcassi.wordtarget.core.CommonTestObjects.*;
import java.awt.List;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Set;

/**
 *
 * @author Matteo
 */
public class WordDatabaseTest {

    @Test
    public void testPopulatingDatabase() {
        WordDatabase db = new WordDatabase();
        db.addWord(cat);
        db.addWords(Arrays.asList(bat, baseball, sport, ports, port));
        assertEquals(Set.of(cat, bat, baseball, sport, ports, port), db.getAllWords());
        Set<Link> batLinks = db.getLinksFor(bat);
        assertEquals(1, batLinks.size());
        assertTrue(batLinks.contains(catBat.reverse()));
        Set<Link> portsLinks = db.getLinksFor(ports);
        assertEquals(2, portsLinks.size());
        assertTrue(portsLinks.contains(new Link(ports, sport, LinkType.Anagram)));
        assertTrue(portsLinks.contains(new Link(ports, port, LinkType.OneLetterAddOrRemove)));
    }

    @Test
    public void testSerializeDatabase() throws IOException {
        String expected = """
Words:
BASEBALL
BAT
CAT
PORT
PORTS
SPORT

Links:
        """;
        StringWriter sw = new StringWriter();
        BufferedWriter writer = new BufferedWriter(sw);
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports, port));
        db.serialize(writer);
        writer.flush();
        assertEquals(expected, sw.toString());
    }

    @Test
    public void testSerializeDatabase1() throws IOException {
        String expected = """
Words:
NURSED
RED SUN
SIGHT
TOWN
VITAMIN A
WON'T

Links:
SIGHT, VITAMIN A, WordAssociation
        """;

        StringWriter sw = new StringWriter();
        BufferedWriter writer = new BufferedWriter(sw);
        WordDatabase db = new WordDatabase();
        db.addWords(Word.of("NURSED", "RED SUN", "SIGHT", "TOWN", "VITAMIN A", "WON'T"));
        db.addLink(new Link(Word.of("VITAMIN A"), Word.of("SIGHT"), LinkType.WordAssociation));
        db.serialize(writer);
        writer.flush();
        assertEquals(expected, sw.toString());
    }

    @Test
    public void testDeserializeDatabase() throws IOException {
        String expected = """
Words:
BASEBALL
BAT
CAT
PORT
PORTS
SPORT

Links:
BASEBALL, BAT, WordAssociation
BASEBALL, SPORT, WordAssociation
        """;

        StringReader sr = new StringReader(expected);
        BufferedReader reader = new BufferedReader(sr);
        WordDatabase db = WordDatabase.deserialize(reader);
        assertEquals(Set.of(baseball, bat, cat, sport, ports, port), db.getAllWords());
        assertTrue(db.containsLink(baseball, bat));
        assertTrue(db.containsLink(bat, baseball));
        assertTrue(db.containsLink(ports, sport));
    }

    @Test
    public void testDeserializeDatabase1() throws IOException {
        String expected = """
Words:
NURSED
RED SUN
SIGHT
TOWN
VITAMIN A
WON'T

Links:
VITAMIN A, SIGHT, WordAssociation
        """;

        StringReader sr = new StringReader(expected);
        BufferedReader reader = new BufferedReader(sr);
        WordDatabase db = WordDatabase.deserialize(reader);
        Word nursed = Word.of("NURSED");
        Word redSun = Word.of("RED SUN");
        Word vitaminA = Word.of("VITAMIN A");
        Word sight = Word.of("SIGHT");
        Word town = Word.of("TOWN");
        Word wont = Word.of("WON'T");
        assertEquals(Set.of(nursed, redSun, vitaminA, sight, wont, town), db.getAllWords());
        assertTrue(db.containsLink(vitaminA, sight));
        assertTrue(db.containsLink(sight, vitaminA));
        assertTrue(db.containsLink(nursed, redSun));
        assertTrue(db.containsLink(redSun, nursed));
        assertTrue(db.containsLink(wont, town));
        assertTrue(db.containsLink(town, wont));
    }

    @Test
    public void testFindChain() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports));
        db.addLink(batBaseball);
        db.addLink(new Link(sport, baseball, LinkType.WordAssociation));
        Chain chain = db.findChain(cat, ports);
        assertEquals(5, chain.size());
        assertEquals(catBat, chain.links().get(0));
        assertEquals(batBaseball, chain.links().get(1));
        assertEquals(baseballSport, chain.links().get(2));
        assertEquals(sportPorts, chain.links().get(3));
    }

    @Test
    public void testImproperWordAdd() {
        WordDatabase db = new WordDatabase();
        db.addWord(cat);
        assertEquals(Set.of(cat), db.getAllWords());
        db.addWord(cat);
        assertEquals(1, db.getAllWords().size());
        assertEquals(Set.of(cat), db.getAllWords());
    }

    @Test
    public void testContainsWord() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball));
        assertTrue(db.containsWord(cat));
        assertTrue(db.containsWord(bat));
        assertTrue(db.containsWord(baseball));
        assertFalse(db.containsWord(ports));
        assertFalse(db.containsWord(sport));
        assertFalse(db.containsWord(null));
    }

    @Test
    public void testContainsLink() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports, port));
        db.addLink(batBaseball);
        assertTrue(db.containsLink(cat, bat));
        assertTrue(db.containsLink(bat, cat));
        assertTrue(db.containsLink(bat, baseball));
        assertTrue(db.containsLink(sport, ports));
        assertFalse(db.containsLink(cat, sport));
        assertFalse(db.containsLink(cat, Word.of("RAT")));
    }

    @Test
    public void testGetLinkBetween() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports, port));
        db.addLink(batBaseball);
        db.addLink(baseballSport);
        assertEquals(catBat, db.getLinkBetween(cat, bat));
        assertEquals(baseballSport, db.getLinkBetween(baseball, sport));
        assertEquals(null, db.getLinkBetween(cat, baseball));
        assertEquals(null, db.getLinkBetween(car, cat));
        assertEquals(null, db.getLinkBetween(null, cat));
    }

    @Test
    public void testRemoveLink() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports, port));
        db.addLink(batBaseball);
        db.addLink(baseballSport);
        assertTrue(db.containsLink(sport, port));
        db.removeLink(sportPort);
        assertFalse(db.containsLink(sport, port));
        assertThrows(IllegalArgumentException.class, () -> {
            db.removeLink(portsParts);
        });
    }
    
    public void testGetLinkForWithExclusions() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball, sport, ports, port));
        db.addLink(batBaseball);
        db.addLink(baseballSport);
        assertEquals(Set.of(catBat.reverse(), batBaseball), db.getLinksFor(bat, Set.of()));
        assertEquals(Set.of(batBaseball), db.getLinksFor(bat, Set.of(cat)));
        assertEquals(Set.of(), db.getLinksFor(bat, Set.of(cat, baseball)));
    }
    
    @Test
    public void testAddFromChain() {
        WordDatabase db = new WordDatabase();
        db.addWords(Arrays.asList(cat, bat, baseball));
        db.addLink(batBaseball);
        db.addLink(baseballSport);

        Chain chain = Chain.of(catBat, batBaseball, baseballSport, sportPorts);
        assertTrue(db.addFromChain(chain));
        assertEquals(Set.of(cat, bat, baseball, sport, ports), db.getAllWords());
        for (Link link : chain.links()) {
            assertTrue(db.containsLink(link));
        }
    }
}
